src/pages/RFOX/components/Stake/StakeConfirm.tsx

```diff
@@ -7,41 +7,404 @@ import {
+  )
+
+  const {
+    data: userBalanceOfCryptoBaseUnit,
+    isSuccess: isUserBalanceOfCryptoBaseUnitSuccess,
+    queryKey: userBalanceOfCryptoBaseUnitQueryKey,
+  } = useReadContract({
+    abi: foxStakingV1Abi,
+    address: RFOX_PROXY_CONTRACT_ADDRESS,
+    functionName: 'balanceOf',
+    args: [getAddress(stakingAssetAccountAddress)],
+    chainId: arbitrum.id,
+    query: {
+      select: data => data.toString(),
+      enabled: Boolean(stakingAsset),
+    },
+  })
+
+  const {
+    data: newContractBalanceOfCryptoBaseUnit,
+    isSuccess: isNewContractBalanceOfCryptoBaseUnitSuccess,
+    queryKey: newContractBalanceOfCryptoBaseUnitQueryKey,
+  } = useReadContract({
+    abi: erc20ABI,
+    address: getAddress(fromAssetId(confirmedQuote.stakingAssetId).assetReference),
+    functionName: 'balanceOf',
+    args: [getAddress(RFOX_PROXY_CONTRACT_ADDRESS)],
+    chainId: arbitrum.id,
+    query: {
+      select: data =>
+        bnOrZero(data.toString()).plus(confirmedQuote.stakingAmountCryptoBaseUnit).toFixed(),
+    },
+  })
+
+  const newShareOfPoolPercentage = useMemo(
+    () =>
+      bnOrZero(confirmedQuote.stakingAmountCryptoBaseUnit)
+        .plus(userBalanceOfCryptoBaseUnit ?? 0)
+        .div(newContractBalanceOfCryptoBaseUnit ?? 0)
+        .toFixed(4),
+    [
+      confirmedQuote.stakingAmountCryptoBaseUnit,
+      newContractBalanceOfCryptoBaseUnit,
+      userBalanceOfCryptoBaseUnit,
+    ],
+  )
+
+  // Approval/Allowance bits
+
+  const { data: allowanceDataCryptoBaseUnit, isLoading: isAllowanceDataLoading } = useAllowance({
+    assetId: stakingAsset?.assetId,
+    spender: RFOX_PROXY_CONTRACT_ADDRESS,
+    from: fromAccountId(confirmedQuote.stakingAssetAccountId).account,
+  })
+
+  const isApprovalRequired = useMemo(
+    () => bnOrZero(allowanceDataCryptoBaseUnit).lt(confirmedQuote.stakingAmountCryptoBaseUnit),
+    [allowanceDataCryptoBaseUnit, confirmedQuote.stakingAmountCryptoBaseUnit],
+  )
+
+  const approvalCallData = useMemo(() => {
+    return encodeFunctionData({
+      abi: erc20ABI,
+      functionName: 'approve',
+      args: [RFOX_PROXY_CONTRACT_ADDRESS, BigInt(confirmedQuote.stakingAmountCryptoBaseUnit)],
+    })
+  }, [confirmedQuote.stakingAmountCryptoBaseUnit])
+
+  const isGetApprovalFeesEnabled = useMemo(
+    () =>
+      Boolean(
+        isApprovalRequired &&
+          stakingAssetAccountNumber !== undefined &&
+          feeAsset &&
+          feeAssetMarketData &&
+          wallet,
+      ),
+    [feeAsset, feeAssetMarketData, isApprovalRequired, stakingAssetAccountNumber, wallet],
+  )
+
+  const {
+    data: approvalFees,
+    isLoading: isGetApprovalFeesLoading,
+    isSuccess: isGetApprovalFeesSuccess,
+  } = useQuery({
+    ...reactQueries.common.evmFees({
+      value: '0',
+      accountNumber: stakingAssetAccountNumber!, // see isGetApprovalFeesEnabled
+      feeAsset: feeAsset!, // see isGetApprovalFeesEnabled
```

Comment:
~~Not sure how that would look like? Are we thinking something a la `isGetApprovalFeesEnabled ? stakingAssetAccountNumber : 0`?~~
~~That's unfortunately always the issue with types and react-query: args are evaluated at component render time, where things may be undefined, but the query we've defined doesn't allow undefined.~~

~~If we wanted to be more accurate, we could have the query itself accept args that are maybe undefined, however that makes us have to check there, which sounds off too since enabled ensures things never run with undefined values.~~

Disregard all the above: as usual with all things react-query, tkdodo wrote about this, and it looks like there's now an even better option: https://tkdodo.eu/blog/react-query-and-type-script

> Since v5.25, You can use a [skipToken](https://tanstack.com/query/v5/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken) to disable a query while having good type safety in queryFn

Captured https://github.com/shapeshift/web/issues/6973 for it, since it's probably not a good idea to bump react-query in this already huge PR!

Reply:
Great find 🙏